Hardhat Developer Workflow Tutorial (#524)
PureStake/moonbeam-docs: tutorials/hardhat-tutorial.md

```
npx hardhat verify --network moonbaseAlpha <CONTRACT-ADDRESS> "{{ networks.moonbase.staking.candidates.address1 }}" "DEPLOYER-ADDRESS"
```

In your terminal you should see the source code for your contract was successfully submitted for verification. If the verification was successful, you should see **Successfully verified contract** and there will be a link to the contract code on [Moonscan for Moonbase Alpha](https://moonbase.moonscan.io/){target=_blank}. If the plugin returns an error, double check that your [API key is configured correctly and that you have specified all necessary parameters in the verification command](/builders/build/eth-api/verify-contracts/etherscan-plugins/){target=_blank}.

![Successful verification using hardhat-etherscan plugin](/images/tutorials/hardhat/hardhat3.png)

## Testing {: #testing }

A robust smart contract development workflow is incomplete without a testing suite. Hardhat has a number of tools that make it easy to write and run tests. In this section, you'll learn the basics of testing your smart contracts and some more advanced techniques such as using fixtures. 

> Hardhat tests are typically written with Mocha and Chai. [Mocha](https://mochajs.org/){target=_blank} is a Javascript testing framework and [Chai](https://www.chaijs.com/){target=_blank} is a BDD/TDD Javascript assertion library. BDD / TDD stands for behavior and test driven development respectively. Effective BDD / TDD necessitates writing your tests *before* writing your smart contract code. The structure of this tutorial doesn't strictly follow these guidelines, but you may wish to adopt these principles in your development workflow. Hardhat previously recommended the Hardhat Waffle plugin but has since [advised migrating to the Hardhat Toolbox](https://hardhat.org/hardhat-runner/docs/guides/migrating-from-hardhat-waffle){target=_blank}.

Reviewer: I would maybe cut out the part about Waffle and just say it's recommended to use the Hardhat Toolbox, which come with Mocha and Chai out of the box (right?)

### Configuring the Test File {: #configuring-the-test-file }

To get started, create a folder called `tests` and a file named `Dao.js`. Then copy and paste the contents below to set up the initial structure of your test file. Be sure to read the comments as they can clarify the purpose of each line. 

``` javascript
const { ethers } = require("hardhat");
require("@nomicfoundation/hardhat-toolbox");

// Import Chai to use its assertion functions here.
const { expect } = require("chai");

// Use `loadFixture` to share common setups (or fixtures) between tests.
const { loadFixture } = require("@nomicfoundation/hardhat-network-helpers");

// Indicate the collator the DAO wants to delegate to
const targetCollator = "{{ networks.moonbase.staking.candidates.address1 }}";
```

### Using Fixtures {: #using-fixtures }

> A fixture is a function that configures the blockchain to a desired state. For example, our initial tests for the StakingDAO will be run after the StakingDAO has been deployed but before any contributions have been made. When a fixture is loaded for the first time, it is executed as normal to properly configure the state of the chain. When called a second time, the fixture is not executed again but rather the state of the network is reset to the point in time immediately after the fixture was initially executed. In other words, this resets state changes prior to the initial execution of the fixture and saves time.

Reviewer: ```suggestion
A fixture is a function that configures the blockchain to a desired state. For example, our initial tests for the staking DAO will be run after the staking DAO has been deployed but before any contributions have been made. When a fixture is loaded for the first time, it is executed as normal to properly configure the state of the chain. When called a second time, the fixture is not executed again but rather the state of the network is reset to the point in time immediately after the fixture was initially executed. In other words, this resets state changes prior to the initial execution of the fixture and saves time.
```

Reviewer: fix all other occurrences... should be "`DelegationDAO`" (to match contract name) or "staking DAO"

The first fixture in the test file represents an "empty" stakingDAO but the second fixture will have one member (besides the DAO admin). Fixtures allow you to quickly switch between different states, simplifying and speeding up the testing process. In this tutorial only two fixtures are specified but it's easy to imagine other states you'd like to test, such as a scenario with 100 DAO members or one with multiple admins of the DAO. 

To define the two fixtures, add the following snippet to your test file:

``` javascript
// `describe` is a Mocha function that allows you to organize your tests. All Mocha
// functions are available in the global scope.
//
// `describe` receives the name of a section of your test suite, and a
// callback. The callback must define the tests of that section. This callback

> // can't be an async function.

Reviewer: I would put the describe info outside of the comments so we can link back to Mocha's docs, but not sure it really matters here 🤷‍♀️

describe("Dao contract", function () {
  async function deployDaoFixture() {

>     // Get the ContractFactory and Signers here.

Reviewer: no punctuation is needed at the end of comments, please fix throughout

    const [deployer] = await ethers.getSigners();
    const delegationDao = await ethers.getContractFactory("DelegationDAO");
    
    // To deploy our StakingDao, you need to call delegationDao.deploy() and await

>     // its deployed() method, which happens once its transaction has been mined.

Reviewer: this references the code, but it should explain what the code does even if it seems obvious. so instead of `delegationDao.deploy()` just say you need to deploy the contract and wait for the deployment transaction to be mined

    const deployedDao = await delegationDao.deploy(targetCollator, deployer.address);

    await deployedDao.deployed();

    // Fixtures can return anything you consider useful for your tests
    return { deployedDao };
  }

    async function deployDaoFixtureWithMembers() {
    const [deployer, member1] = await ethers.getSigners();

    const delegationDao = await ethers.getContractFactory("DelegationDAO");
    const deployedDao = await delegationDao.deploy(targetCollator, deployer.address);

    await deployedDao.deployed();
    await deployedDao.grant_member(member1.address);

>     return { deployedDao};

Reviewer: add comments

>   }

Reviewer: this is missing `});` at the end to enclose the `describe` function

```

### Writing your First Test Cases {: #writing-your-first-test-cases }

This tutorial deviates from the [test driven development philosophy](https://en.wikipedia.org/wiki/Test-driven_development){target=_blank} of writing your tests prior to your code. A few test cases will be cherry picked for demonstration purposes. 

> First, you'll create a subsection called `deployment` to keep the test file organized. This isn't required but may be helpful for organization purposes. Next you'll define your first test case by using the `it` mocha function. A description that clearly indicates what the test is doing is provided as well as indication that the function is async. This first test is simply checking to see that the StakingDAO is correctly storing the address of the target collator.

Reviewer: should be `Deployment` to match the code, or update code to match this

Reviewer: Mocha should be capitalized

Next, a fixture is loaded - in this case, the "empty" stakingDAO fixture. Fixtures are easily interchangeable within your test cases. 

> 

Reviewer: I would explicitly state that this snippet needs to be added to the describe dao contract function

```javascript
// You can nest describe calls to create subsections.
describe("Deployment", function () {
    // `it` is another Mocha function. This is the one you use to define each
    // of your tests. It receives the test name, and a callback function.
    //
    // If the callback function is async, Mocha will `await` it.
    it("Should have correct target collator", async function () {
      // We use loadFixture to setup our environment, and then assert that
      // things went well
      const { deployedDao, deployer } = await loadFixture(deployDaoFixture);

      // `expect` receives a value and wraps it in an assertion object. These
      // objects have a lot of utility methods to assert values.
      expect(await deployedDao.target()).to.equal(targetCollator);

> });

Reviewer: this is missing `});` at the end to enclose the describe function

```

Let's add another test case. When a StakingDAO is launched, it shouldn't have any funds. This test verifies that is indeed the case. Go ahead and add the following test case to your `Dao.js` file:

```javascript

> it("The DAO should initially have 0 funds in it", async function () {

Reviewer: ```suggestion
it("should initially have 0 funds in the DAO", async function () {
```

Reviewer: add comments here and where missing throughout the guide

	const { deployedDao, deployer } = await loadFixture(deployDaoFixture);
	expect(await deployedDao.totalStake()).to.equal(0);
});

> ```

Reviewer: I ran these tests and they both failed. So I tried to look up `loadFixture` in hardhats docs and it only shows up on the hardhat-waffle plugin page so I am a bit confused: https://hardhat.org/hardhat-runner/plugins/nomiclabs-hardhat-waffle#environment-extensions

The error I got when running the tests:

```
     OnlyHardhatNetworkError: This helper can only be used with Hardhat Network. You are connected to 'moonbase'.
      at checkIfHardhatNetwork (/Users/es/Documents/work/kevin-hardhat-tutoral/hardhat/node_modules/@nomicfoundation/hardhat-network-helpers/src/utils.ts:28:11)
      at getHardhatProvider (/Users/es/Documents/work/kevin-hardhat-tutoral/hardhat/node_modules/@nomicfoundation/hardhat-network-helpers/src/utils.ts:39:9)
      at takeSnapshot (/Users/es/Documents/work/kevin-hardhat-tutoral/hardhat/node_modules/@nomicfoundation/hardhat-network-helpers/src/helpers/takeSnapshot.ts:20:20)
      at loadFixture (/Users/es/Documents/work/kevin-hardhat-tutoral/hardhat/node_modules/@nomicfoundation/hardhat-network-helpers/src/loadFixture.ts:60:22)
      at Context.<anonymous> (/Users/es/Documents/work/kevin-hardhat-tutoral/hardhat/tests/Dao.js:65:47)
      ```

Author: My bad, yup fixtures are only for hardhat network so I removed any references to fixtures.

### Function Reverts {: #function-reverts }

> To this point the two cases implemented have been simple but important. Now, you'll implement a more complex test case that differs in its architecture. In prior examples you've verified that a function returns an expected value. In this one, you'll be verifying that a function reverts. You'll also change the address of the caller to test an admin-only function. 

Reviewer: ```suggestion
To this point, the two cases implemented have been simple but important. Now, you'll implement a more complex test case that differs in its architecture. In prior examples, you've verified that a function returns an expected value. In this one, you'll be verifying that a function reverts. You'll also change the address of the caller to test an admin-only function. 
```

> In the [StakingDAO contract](https://github.com/PureStake/moonbeam-intro-course-resources/blob/main/delegation-dao-lesson-one/DelegationDAO.sol){target=_blank}, only admins are authorized to add new members to the DAO. One could write a test that checks to see if the admin is authorized to add new members but perhaps a more important test is to ensure that *non-admins* can't add new members. To run this test case under a different account, you're going to ask for another address when you call `ethers.getSigners()` and specify the caller in the assertion with `connect(otherAddress)`. Finally, after the function call to be tested you'll append `.to.be.reverted;` to indicate that the test case is successful if the function reverts. And if it doesn't revert it's a failed test! 

Reviewer: so do you need to set up a different account in the hardhat config file to do this?

Author: not strictly necessary

```javascript

> it("Non-admins should not be able to grant membership", async function () {

Reviewer: test descriptions should read like a sentence, so "it should not allow non-admints to grant membership"
```suggestion
it("should not allow non-admins to grant membership", async function () {
```

	const { deployedDao, deployer } = await loadFixture(deployDaoFixture);
	const [account1, otherAddress] = await ethers.getSigners();
	await expect(deployedDao.connect(otherAddress).grant_member("0x0000000000000000000000000000000000000000")).to.be.reverted;
});
```

### Swapping Fixtures {: #swapping-fixtures }

For this example, you'll load another fixture - the one that has a DAO deployed with a member. Then, you'll check to verify whether that member can call the `check_free_balance()` function of StakingDAO, which has an access modifier such that only members can access it. 

```javascript
it("DAO members should be able to access member only functions", async function () {
	const { deployedDao, deployer } = await loadFixture(deployDaoFixtureWithMembers);
    const [account1, member1] = await ethers.getSigners();
    expect(await deployedDao.connect(member1).check_free_balance()).to.equal(0);
	});
  });
});
```

> And that's it! You're now ready to run your tests! Since this is the last test case completing the test file, a series of closing brackets are added to the prior code snippet.  

Reviewer: ah ok I see now you mentioned adding in the closing brackets, but I feel like it's better to have them in each of the snippets in case a user wants to run the tests as they add them 🤷‍♀️

### Running your Tests {: #running-your-tests }

If you've followed all of the prior sections, your `Dao.js` test file should be all set to go. Otherwise, you can copy the complete snippet below into your `Dao.js` test file. Comments have been removed for code readability, but you can refer to the prior sections for details on each step including comments.

> 

Reviewer: To shorten this, I would add the complete script as a `.snippet/code` and link to the raw version on github

  	  expect(await deployedDao.totalStake()).to.equal(0);
});

    it("Non-admins should not be able to grant membership", async function () {
  	  const { deployedDao, deployer } = await loadFixture(deployDaoFixture);
  	  const [account1, otherAddress] = await ethers.getSigners();
 	  await expect(deployedDao.connect(otherAddress).grant_member("0x0000000000000000000000000000000000000000")).to.be.reverted;
});

    it("DAO members should be able to access member only functions", async function () {
      const { deployedDao, deployer } = await loadFixture(deployDaoFixtureWithMembers);
      const [account1, member1] = await ethers.getSigners();
      expect(await deployedDao.connect(member1).check_free_balance()).to.equal(0);
    });
  });
});
```

You can run your tests with the following command: 
```npx hardhat test tests/Dao.js```

If everything was set up correctly, you should see output like the following: 

![Hardhat Run Tests](/images/tutorials/hardhat/hardhat4.png)

And that's it! We covered a lot of ground in this tutorial but there's more resources available if you'd like to go deeper, including the following:

> - [Hardhat Guide to Testing Contracts](https://hardhat.org/hardhat-runner/docs/guides/test-contracts){target=_blank}
> - [Writing Tests and Scripts](https://hardhat.org/hardhat-runner/docs/guides/tasks-and-scripts){target=_blank}

Reviewer: ```suggestion
- [Hardhat guide to Testing Contracts](https://hardhat.org/hardhat-runner/docs/guides/test-contracts){target=_blank}
- [Writing tasks and scripts](https://hardhat.org/hardhat-runner/docs/guides/tasks-and-scripts){target=_blank}
```

Reviewer: capitalization is based on titles that hardhat uses

```javascript
it("should only allow members to access member-only functions", async function () {
  const { deployedDao } = await deployDao();

  // We ask ethers for two accounts back this time
  const [deployer, member1] = await ethers.getSigners();

  // This test will succeed if the DAO member can call the member-only function.
  // We use connect here to call the function from the account of the new member
  expect(await deployedDao.connect(member1).check_free_balance()).to.equal(0);
});
```

And that's it! You're now ready to run your tests!

### Running your Tests {: #running-your-tests }

If you've followed all of the prior sections, your [`Dao.js`](https://raw.githubusercontent.com/PureStake/moonbeam-intro-course-resources/main/delegation-dao-lesson-one/Dao.js){target=_blank} test file should be all set to go. Otherwise, you can copy the [complete snippet from GitHub](https://github.com/PureStake/moonbeam-docs/blob/master/.snippets/code/hardhat/dao-js-test-file.js) into your `Dao.js` test file.

> Since our test cases encompass mostly configuration and setup of the staking DAO and don't involve actual delegation actions, we'll be running our tests on a Moonbeam dev node (local node). If you wrote different test cases that involved delegation to collators (e.g. when the DAO is funded sufficiently) you'll want to utilize Moonbase Alpha instead. You can use the flag `--network moonbase` to run the tests using Moonbase Alpha. In that case, be sure that your deployer address is sufficiently funded with DEV tokens. You can also omit the `--network` flag in which case the tests will run against a local hardhat network. 

Reviewer: update this paragraph as they can do test cases on a development node if they delegate Alice

In your terminal you should see the source code for your contract was successfully submitted for verification. If the verification was successful, you should see **Successfully verified contract** and there will be a link to the contract code on [Moonscan for Moonbase Alpha](https://moonbase.moonscan.io/){target=_blank}. If the plugin returns an error, double check that your API key is configured correctly and that you have specified all necessary parameters in the verification command. You can refer to the [guide to the Hardhat Etherscan plugin](/builders/build/eth-api/verify-contracts/etherscan-plugins/){target=_blank} for more information.

![Verify contracts on Moonbase Alpha using the Hardhat Etherscan plugin.](/images/tutorials/hardhat/hardhat-6.png)

> ## Deploying to Production on Moonbeam Mainnet {: #deploying-to-production-on-moonbeam-mainnet }

Reviewer: mention that they'll need to update the hardhat.config.js to add moonbeam as a network before using the `--network moonbeam` command
